### ISSUE TITLE
System is structurally singular and cannot be handled because the number of unassigned equations is larger than the number of states.

### ISSUE THREAD

**Issue by rfranke** [OpenModelica/OpenModelica#7818]
A couple of PowerSystems examples fail with this message. I just discovered that the `PumpingSystem` example is suffering from the same bug in MSL 3.2.3. Interestingly it works in `OpenModelica/testsuite/simulation/libraries/msl32`.
This might give some hints and it should also raise the importance of a solution. 

See https://libraries.openmodelica.org/branches/master/Modelica_3.2.3/files/Modelica_3.2.3_Modelica.Fluid.Examples.PumpingSystem.err
```
Error: Internal error IndexReduction.pantelidesIndexReduction failed! System is structurally singular and cannot be handled because the number of unassigned equations is larger than the number of states. Use -d=bltdump to get more information.
Notification: Performance of <failed> postOpt replaceHomotopyWithSimplified (initialization): time 0.00295/1.379, allocations: 0.7657 MB / 0.6069 GB, free: 10.67 MB / 442.7 MB
Error: post-optimization module replaceHomotopyWithSimplified (initialization) failed.
Error: No system for the symbolic initialization was generated
```

**Comment by casella** [comment]
Possibly related to [#6200](https://trac.openmodelica.org/OpenModelica/ticket/6200)

**Comment by casella** [comment]
Regarding msl 3.2, I guess homotopy was added afterwards, that may explain what goes wrong.

It is possible that the simplifed equation changes the index of the problem. I'm not sure if this can be handled and how, I need to think about it.

**Comment by kabdelhak** [comment]
This issue seems to be resolved with PRs #7853 and #7875:
```
""
record SimulationResult
    resultFile = "/home/kab/Repo/FH/promotion/modelicaLibs/ticketTests/7818_pantelides_fail/Modelica.Fluid.Examples.PumpingSystem_res.mat",
    simulationOptions = "startTime = 0.0, stopTime = 2000.0, numberOfIntervals = 5000, tolerance = 1e-06, method = 'dassl', fileNamePrefix = 'Modelica.Fluid.Examples.PumpingSystem', options = '', outputFormat = 'mat', variableFilter = '.*', cflags = '', simflags = ''",
    messages = "assert            | warning | The following assertion has been violated at time 0.000000
|                 | |       | pipe.flowModel.m_flows[1] >= 0.0 and pipe.flowModel.m_flows[1] <= 100000.0
assert            | warning | Variable violating min/max constraint: 0.0 <= pipe.flowModel.m_flows[1] <= 100000.0, has value: -141.424
LOG_SUCCESS       | info    | The initialization finished successfully with 3 homotopy steps.
assert            | warning | The following assertion has been violated at time 200.000000
|                 | |       | valveOpening.y >= 0.0 and valveOpening.y <= 1.0
assert            | warning | Variable violating min/max constraint: 0.0 <= valveOpening.y <= 1.0, has value: 1
LOG_SUCCESS       | info    | The simulation finished successfully.
",
    timeFrontend = 0.413479731,
    timeBackend = 0.387893598,
    timeSimCode = 0.032612952,
    timeTemplates = 0.16161118,
    timeCompile = 5.518726765,
    timeSimulation = 2.772669733,
    timeTotal = 9.287098070000001
end SimulationResult;
"[/home/kab/workspace/OM_master/build/lib/omlibrary/Modelica 4.0.0/Fluid/Sources.mo:730:36-731:54:writable] Warning: 'each' used when modifying non-array element m_flow.
[/home/kab/workspace/OM_master/build/lib/omlibrary/Modelica 4.0.0/Fluid/Sources.mo:728:36-729:54:writable] Warning: 'each' used when modifying non-array element m_flow.
[/home/kab/workspace/OM_master/build/lib/omlibrary/Modelica 4.0.0/Fluid/Sources.mo:730:36-731:54:writable] Warning: 'each' used when modifying non-array element m_flow.
[/home/kab/workspace/OM_master/build/lib/omlibrary/Modelica 4.0.0/Fluid/Sources.mo:728:36-729:54:writable] Warning: 'each' used when modifying non-array element m_flow.
Warning: The model contains alias variables with redundant start and/or conflicting nominal values. It is recommended to resolve the conflicts, because otherwise the system could be hard to solve. To print the conflicting alias sets and the chosen candidates please use -d=aliasConflicts.
"
```
Can someone verify the results?

**Comment by rfranke** [comment]
The current trunk version shows no improvement for me. Your logs says MSL 4.0.0. There it is working in the nightly tests too. Please try MSL 3.2.3.

**Comment by rfranke** [comment]
Here is another example that works with a slight variation:
```
PowerSystems.Examples.Introductory.SimulationSteadyInitial
```
raises the same error [System is structurally singular and cannot be handled because the number of unassigned equations is larger than the number of states.](https://libraries.openmodelica.org/branches/master/PowerSystems_latest/files/PowerSystems_latest_PowerSystems.Examples.Introductory.SimulationSteadyInitial.err)

The same example works fluently if not the initial equations, but the regular equations are formulated in steady-state:
```
PowerSystems.Examples.Introductory.SimulationSteadyState
```
The two models are [identical except for one parameter](https://github.com/modelica-3rdparty/PowerSystems/blob/master/PowerSystems/Examples/Introductory.mo):
```modelica
  model SimulationSteadyInitial
    extends SimulationFixedInitial(system(dynType=PowerSystems.Types.Dynamics.SteadyInitial));
  end SimulationSteadyInitial;

  model SimulationSteadyState "Steady-state simulation"
    extends SimulationFixedInitial(system(dynType=PowerSystems.Types.Dynamics.SteadyState));
  end SimulationSteadyState;
```
This switches the equations in the used components:
```modelica
  model RXline "RX transmission line, 3-phase dq0"
    extends Ports.Port_pn;
    extends Partials.RXlineBase(final ne=1);

    PS.Voltage[PS.n] v(start = v_start);
    PS.Current[PS.n] i(start = i_start);
  protected
    SI.AngularFrequency[2] omega;

  initial equation
    if dynType == Types.Dynamics.SteadyInitial then
      der(i) = omega[1]*j(i);
    elseif dynType == Types.Dynamics.FixedInitial then
      i = i_start;
    end if;

  equation
    omega = der(term_p.theta);
    v = term_p.v - term_n.v;
    i = term_p.i;

    if dynType <> Types.Dynamics.SteadyState then
      PS.map({L,L,L0}).*der(i) + omega[2]*L*j(i) + R*i = v;
    else
      omega[2]*L*j(i) + R*i = v;
    end if;
end RXline;
```

**Comment by rfranke** [comment]
Here are examples from [Buildings 7.0](https://libraries.openmodelica.org/branches/master/Buildings_maint.7.0.x/Buildings_maint.7.0.x.html) suffering from the same problem. These are 18 out of 76 Buildings models currently failing in the backend:
```
Buildings.Air.Systems.SingleZone.VAV.Examples.ChillerDXHeatingEconomizer
Buildings.Air.Systems.SingleZone.VAV.Examples.Guideline36
Buildings.Examples.ScalableBenchmarks.BuildingVAV.Examples.OneFloor_OneZone
Buildings.Examples.VAVReheat.ASHRAE2006
Buildings.Examples.VAVReheat.Guideline36
Buildings.Examples.VAVReheat.Validation.Guideline36SteadyState
Buildings.Experimental.DistrictHeatingCooling.Examples.HeatingCoolingHotWater3Clusters
Buildings.Experimental.DistrictHeatingCooling.Validation.HeatingCoolingHotWaterSmall
Buildings.Fluid.HeatExchangers.BaseClasses.Examples.HexElementLatentLoop
Buildings.Fluid.HeatExchangers.DXCoils.AirCooled.Examples.MultiStage
Buildings.Fluid.HeatExchangers.DXCoils.AirCooled.Examples.VariableSpeed
Buildings.Fluid.HeatExchangers.Examples.AirHeater_u
Buildings.Fluid.HeatExchangers.RadiantSlabs.Examples.SingleCircuitMultipleCircuitFiniteDifference
Buildings.Fluid.HeatExchangers.RadiantSlabs.Examples.StepResponseEpsilonNTU
Buildings.Fluid.HeatExchangers.RadiantSlabs.Examples.StepResponseFiniteDifference
Buildings.Fluid.HeatExchangers.Validation.HeaterCooler_u
Buildings.ThermalZones.ReducedOrder.Examples.SimpleRoomFourElementsTraceSubstance
Buildings.Utilities.Plotters.Examples.SingleZoneVAV
```
Overall it appears that only few backend errors repeat again and again for the 76 failing models.

**Comment by casella** [comment]
@rfranke, we may have a solution in the pipeline as discussed in [#6200](https://trac.openmodelica.org/OpenModelica/ticket/6200), see PR #7903. Please follow that.

**Comment by casella** [comment]
@rfranke, 

> Here are examples from [Buildings 7.0](https://libraries.openmodelica.org/branches/master/Buildings_maint.7.0.x/Buildings_maint.7.0.x.html) suffering from the same problem. These are 18 out of 76 Buildings models currently failing in the backend:
> 
> ```
> Buildings.Air.Systems.SingleZone.VAV.Examples.ChillerDXHeatingEconomizer
> Buildings.Air.Systems.SingleZone.VAV.Examples.Guideline36
> Buildings.Examples.ScalableBenchmarks.BuildingVAV.Examples.OneFloor_OneZone
> Buildings.Examples.VAVReheat.ASHRAE2006
> Buildings.Examples.VAVReheat.Guideline36
> Buildings.Examples.VAVReheat.Validation.Guideline36SteadyState
> Buildings.Experimental.DistrictHeatingCooling.Examples.HeatingCoolingHotWater3Clusters
> Buildings.Experimental.DistrictHeatingCooling.Validation.HeatingCoolingHotWaterSmall
> Buildings.Fluid.HeatExchangers.BaseClasses.Examples.HexElementLatentLoop
> Buildings.Fluid.HeatExchangers.DXCoils.AirCooled.Examples.MultiStage
> Buildings.Fluid.HeatExchangers.DXCoils.AirCooled.Examples.VariableSpeed
> Buildings.Fluid.HeatExchangers.Examples.AirHeater_u
> Buildings.Fluid.HeatExchangers.RadiantSlabs.Examples.SingleCircuitMultipleCircuitFiniteDifference
> Buildings.Fluid.HeatExchangers.RadiantSlabs.Examples.StepResponseEpsilonNTU
> Buildings.Fluid.HeatExchangers.RadiantSlabs.Examples.StepResponseFiniteDifference
> Buildings.Fluid.HeatExchangers.Validation.HeaterCooler_u
> Buildings.ThermalZones.ReducedOrder.Examples.SimpleRoomFourElementsTraceSubstance
> Buildings.Utilities.Plotters.Examples.SingleZoneVAV
> ```

After #7903 was merged in, all these examples pass the backend phase. Many are successful, some fail at runtime because of numerical issues, e.g. [Buildings.Utilities.Plotters.Examples.SingleZoneVAV](https://libraries.openmodelica.org/branches/master/Buildings_maint.7.0.x/files/Buildings_maint.7.0.x_Buildings.Utilities.Plotters.Examples.SingleZoneVAV.sim), which fails giving:
```
LOG_SUCCESS       | info    | The initialization finished successfully with 3 homotopy steps.
assert            | debug   | division by zero at time 15552000, (a=0) / (b=0), where divisor b expression is: hvac.cooCoi.m1_flow + hvac.ideVal.preMasFlo.m_flow
assert            | debug   | division by zero at time 15552000, (a=0) / (b=0), where divisor b expression is: hvac.cooCoi.m1_flow + hvac.ideVal.preMasFlo.m_flow
assert            | debug   | division by zero at time 15552000, (a=0) / (b=0), where divisor b expression is: hvac.cooCoi.m1_flow + hvac.ideVal.preMasFlo.m_flow
```
I'm not sure if this runtime error is somewhat correlated to the previous issue not being fully resolved, but that is another story anyway.

**Comment by casella** [comment]
Unfortunately, the problem with [SimulationSteadyInitial](https://libraries.openmodelica.org/branches/master/PowerSystems_latest/files/PowerSystems_latest_PowerSystems.Examples.Introductory.SimulationSteadyInitial.err) is still there. It may be related to #6196, that we'll have to address next year to get to 100% Buildings success rate.
